astlib/pprintast.ml

```diff
@@ -21,13 +21,72 @@
 (* Extensive Rewrite: Hongbo Zhang: University of Pennsylvania *)
 (* TODO more fine-grained precedence pretty-printing *)
 
-open Import
+open Ast_412
 open Asttypes
 open Format
 open Location
 open Longident
 open Parsetree
-open Ast_helper
+
+let varify_type_constructors var_names t =
+  let check_variable vl loc v =
+    if List.mem v vl then
+      Location.raise_errorf ~loc "variable in scope syntax error: %s" v in
+  let var_names = List.map (fun v -> v.txt) var_names in
+  let rec loop t =
+    let desc =
+      match t.ptyp_desc with
+      | Ptyp_any -> Ptyp_any
+      | Ptyp_var x ->
+        check_variable var_names t.ptyp_loc x;
+        Ptyp_var x
+      | Ptyp_arrow (label,core_type,core_type') ->
+        Ptyp_arrow(label, loop core_type, loop core_type')
+      | Ptyp_tuple lst -> Ptyp_tuple (List.map loop lst)
+      | Ptyp_constr( { txt = Longident.Lident s}, [])
+        when List.mem s var_names ->
+        Ptyp_var s
+      | Ptyp_constr(longident, lst) ->
+        Ptyp_constr(longident, List.map loop lst)
+      | Ptyp_object (lst, o) ->
+        Ptyp_object (List.map loop_object_field lst, o)
+      | Ptyp_class (longident, lst) ->
+        Ptyp_class (longident, List.map loop lst)
+      | Ptyp_alias(core_type, string) ->
+        check_variable var_names t.ptyp_loc string;
+        Ptyp_alias(loop core_type, string)
+      | Ptyp_variant(row_field_list, flag, lbl_lst_option) ->
+        Ptyp_variant(List.map loop_row_field row_field_list,
+                     flag, lbl_lst_option)
+      | Ptyp_poly(string_lst, core_type) ->
+        List.iter (fun v ->
+          check_variable var_names t.ptyp_loc v.txt) string_lst;
+        Ptyp_poly(string_lst, loop core_type)
+      | Ptyp_package(longident,lst) ->
+        Ptyp_package(longident,List.map (fun (n,typ) -> (n,loop typ) ) lst)
+      | Ptyp_extension (s, arg) ->
+        Ptyp_extension (s, arg)
+    in
+    {t with ptyp_desc = desc}
+  and loop_row_field field =
+    let prf_desc = match field.prf_desc with
+      | Rtag(label,flag,lst) ->
+        Rtag(label,flag,List.map loop lst)
+      | Rinherit t ->
+        Rinherit (loop t)
+    in
+    { field with prf_desc; }
+  and loop_object_field field =
+    let pof_desc = match field.pof_desc with
+      | Otag(label, t) ->
+        Otag(label, loop t)
+      | Oinherit t ->
+        Oinherit (loop t)
+    in
+    { field with pof_desc; }
+  in
+  loop t
```

Comment:
I've copied this from `Ast_helper`. Another option would be to use the compiler `Ast_helper.varify_type_constructors` by adding it to `Astlib`. But I'm avoiding that,  since we have our own `Ast_helper` inside `ppxlib` and don't need the compiler `Ast_helper` anywhere but here and this will be removed at some point in the future.